File: go/dnd-character/dnd_character.go
package dndcharacter

import "math/rand"
import "slices"
import "math"

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

func throwD6() int {
	return rand.Intn(6) + 1
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor(float64(score-10) / 2))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	var dice []int
	for i := 0; i < 4; i++ {
		dice = append(dice, throwD6())
	}
	slices.Sort(dice)
	s := 0
	for _, elem := range dice[1:] {
		s += elem
	}
	return s

}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	ch := Character{
		Strength:     Ability(),
		Dexterity:    Ability(),
		Constitution: Ability(),
		Intelligence: Ability(),
		Wisdom:       Ability(),
		Charisma:     Ability(),
	}
	ch.Hitpoints = 10 + Modifier(ch.Constitution)
	return ch
}

File: kotlin/dnd-character/src/main/kotlin/DndCharacter.kt
import kotlin.math.floor
import kotlin.random.Random

class DndCharacter {
    val strength: Int = rollDice()
    val dexterity: Int = rollDice()
    val constitution: Int = rollDice()
    val intelligence: Int = rollDice()
    val wisdom: Int = rollDice()
    val charisma: Int = rollDice()

    val hitpoints: Int = 10 + DndCharacter.modifier(constitution)

    companion object {
        fun ability(): Int {
            val obj = DndCharacter()

            val abilities =
                listOf(
                    obj.strength,
                    obj.dexterity,
                    obj.constitution,
                    obj.intelligence,
                    obj.wisdom,
                    obj.charisma,
                )
            return abilities.random()
        }

        fun modifier(score: Int): Int {
            val difference = score - 10
            val modifierNotRounded = difference / 2.0
            return floor(modifierNotRounded).toInt()
        }

        private fun rollDice(): Int {
            val list = List(4) { Random.nextInt(6) }
            return list.sorted().take(3).sum()
        }
    }
}
